File: 프로그래머스/2/42626. 더 맵게/더 맵게.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(vector<int> scoville, int K) {
    int answer = 0;
    
    priority_queue<int, vector<int>, greater<>> pq;
    for(const auto& s : scoville) {
        pq.push(s);
    }
    
    int cnt = 0;
    while(pq.top() < K && !pq.empty()) {
        
        
        int least = pq.top();
        pq.pop();
        
        if(pq.empty()) {
            return -1;
        }
        int secondLeast = pq.top();
        pq.pop();
        
        int mix = least + secondLeast * 2;
        pq.push(mix);
        
        cnt++;
    }
    answer = cnt;
    return answer;
}

File: 프로그래머스/2/42626. 더 맵게/더 맵게.java
import java.util.PriorityQueue;

class Solution {
    public int solution(int[] scoville, int K) {
        int answer = 0;
        
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        
        for(int i : scoville) {
            pq.add(i);
        }
        
        while(pq.size() >= 2 && pq.peek() < K) {
            int first = pq.poll();
            int second = pq.poll();
            int mix = first + second * 2;
            
            answer++;
            
            pq.add(mix);  
        }
        
        if(pq.peek() < K) {
            answer = -1;
        }
        
        return answer;
    }
}